File: cloudSync.js
const https = require('https');

class CloudSync {
  constructor(apiUrl = 'https://pixelpantry.alfelfriki.tech') {
    this.apiUrl = apiUrl;
    this.syncInProgress = false;
  }

  // Make HTTP request
  async request(endpoint, method = 'GET', data = null) {
    return new Promise((resolve, reject) => {
      const url = new URL(endpoint, this.apiUrl);
      
      const options = {
        hostname: url.hostname,
        path: url.pathname + url.search,
        method: method,
        headers: {
          'Content-Type': 'application/json',
        }
      };

      if (data) {
        const body = JSON.stringify(data);
        options.headers['Content-Length'] = Buffer.byteLength(body);
      }

      const req = https.request(options, (res) => {
        let responseData = '';

        res.on('data', (chunk) => {
          responseData += chunk;
        });

        res.on('end', () => {
          try {
            const parsed = JSON.parse(responseData);
            if (res.statusCode >= 200 && res.statusCode < 300) {
              resolve(parsed);
            } else {
              reject(new Error(`HTTP ${res.statusCode}: ${parsed.error || responseData}`));
            }
          } catch (error) {
            reject(new Error(`Failed to parse response: ${responseData}`));
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      if (data) {
        req.write(JSON.stringify(data));
      }

      req.end();
    });
  }

  // Sync categories to cloud
  async syncCategoriesToCloud(categories) {
    try {
      for (const category of categories) {
        await this.request('/api/categories', 'POST', category);
      }
      return { success: true };
    } catch (error) {
      console.error('Error syncing categories:', error);
      return { success: false, error: error.message };
    }
  }

  // Sync subcategories to cloud
  async syncSubcategoriesToCloud(subcategories) {
    try {
      for (const subcategory of subcategories) {
        await this.request('/api/subcategories', 'POST', subcategory);
      }
      return { success: true };
    } catch (error) {
      console.error('Error syncing subcategories:', error);
      return { success: false, error: error.message };
    }
  }

  // Sync items to cloud
  async syncItemsToCloud(items) {
    try {
      for (const item of items) {
        await this.request('/api/items', 'POST', {
          id: item.id,
          name: item.name,
          category_id: item.category_id,
          subcategory_id: item.subcategory_id,
          quantity: item.quantity,
          unit: item.unit,
          low_stock_threshold: item.low_stock_threshold,
          created_at: item.created_at
        });
      }
      return { success: true };
    } catch (error) {
      console.error('Error syncing items:', error);
      return { success: false, error: error.message };
    }
  }

  // Get all data from cloud
  async getCloudData() {
    try {
      const [categories, subcategories, items] = await Promise.all([
        this.request('/api/categories', 'GET'),
        this.request('/api/subcategories', 'GET'),
        this.request('/api/items', 'GET')
      ]);

      return {
        success: true,
        data: {
          categories: categories.results || [],
          subcategories: subcategories.results || [],
          items: items.results || []
        }
      };
    } catch (error) {
      console.error('Error getting cloud data:', error);
      return { success: false, error: error.message };
    }
  }

  // Push local data to cloud
  async pushToCloud(localData) {
    if (this.syncInProgress) {
      return { success: false, error: 'Sync already in progress' };
    }

    this.syncInProgress = true;

    try {
      // Clear cloud data first
      await this.request('/api/clear', 'POST');

      // Push categories
      const categoriesResult = await this.syncCategoriesToCloud(localData.categories);
      if (!categoriesResult.success) {
        throw new Error('Failed to sync categories');
      }

      // Push subcategories
      const subcategoriesResult = await this.syncSubcategoriesToCloud(localData.subcategories);
      if (!subcategoriesResult.success) {
        throw new Error('Failed to sync subcategories');
      }

      // Push items
      const itemsResult = await this.syncItemsToCloud(localData.items);
      if (!itemsResult.success) {
        throw new Error('Failed to sync items');
      }

      this.syncInProgress = false;
      return { success: true, message: 'Data pushed to cloud successfully' };
    } catch (error) {
      this.syncInProgress = false;
      console.error('Error pushing to cloud:', error);
      return { success: false, error: error.message };
    }
  }

  // Pull data from cloud to local
  async pullFromCloud() {
    if (this.syncInProgress) {
      return { success: false, error: 'Sync already in progress' };
    }

    this.syncInProgress = true;

    try {
      const result = await this.getCloudData();
      this.syncInProgress = false;

      if (result.success) {
        return {
          success: true,
          data: result.data,
          message: 'Data pulled from cloud successfully'
        };
      } else {
        return result;
      }
    } catch (error) {
      this.syncInProgress = false;
      console.error('Error pulling from cloud:', error);
      return { success: false, error: error.message };
    }
  }

  // Test connection
  async testConnection() {
    try {
      await this.request('/api/health', 'GET');
      return { success: true, message: 'Connected to cloud' };
    } catch (error) {
      return { success: false, error: error.message };
    }
  }
}

module.exports = CloudSync;

File: src-tauri/src/cloud_sync.rs
use crate::models::{Category, CloudItem, Subcategory};
use reqwest::blocking::Client;
use reqwest::Method;
use serde::de::DeserializeOwned;
use serde_json::Value;

#[derive(Clone)]
pub struct CloudSync {
    api_url: String,
    api_key: Option<String>,
    client: Client,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct CloudResponse {
    pub success: bool,
    pub message: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct CloudData {
    pub subcategories: Vec<Subcategory>,
    pub items: Vec<CloudItem>,
}

impl CloudSync {
    pub fn new(api_url: String, api_key: Option<String>) -> Result<Self, String> {
        let client = Client::builder()
            .user_agent("PixelPantry")
            .build()
            .map_err(|err| err.to_string())?;
        Ok(Self {
            api_url: api_url.trim_end_matches('/').to_string(),
            api_key,
            client,
        })
    }

    pub fn test_connection(&self) -> CloudResponse {
        match self.request_value("/api/health", Method::GET, None) {
            Ok(_) => CloudResponse {
                success: true,
                message: Some("Connected to cloud".to_string()),
                error: None,
            },
            Err(error) => CloudResponse {
                success: false,
                message: None,
                error: Some(error),
            },
        }
    }

    pub fn push_to_cloud(
        &self,
        categories: &[Category],
        subcategories: &[Subcategory],
        items: &[CloudItem],
    ) -> CloudResponse {
        if let Err(error) = self.request_value("/api/clear", Method::POST, None) {
            return CloudResponse {
                success: false,
                message: None,
                error: Some(error),
            };
        }

        if let Err(error) = self.sync_collection("/api/categories", categories) {
            return CloudResponse {
                success: false,
                message: None,
                error: Some(error),
            };
        }

        if let Err(error) = self.sync_collection("/api/subcategories", subcategories) {
            return CloudResponse {
                success: false,
                message: None,
                error: Some(error),
            };
        }

        if let Err(error) = self.sync_collection("/api/items", items) {
            return CloudResponse {
                success: false,
                message: None,
                error: Some(error),
            };
        }

        CloudResponse {
            success: true,
            message: Some("Data pushed to cloud successfully".to_string()),
            error: None,
        }
    }

    pub fn pull_from_cloud(&self) -> Result<CloudData, String> {
        let _categories: Vec<Category> = self.request_collection("/api/categories")?;
        let subcategories: Vec<Subcategory> = self.request_collection("/api/subcategories")?;
        let items: Vec<CloudItem> = self.request_collection("/api/items")?;
        Ok(CloudData {
            subcategories,
            items,
        })
    }

    fn request_collection<T: DeserializeOwned>(&self, endpoint: &str) -> Result<Vec<T>, String> {
        let value = self.request_value(endpoint, Method::GET, None)?;
        let results = value.get("results").cloned().unwrap_or(Value::Array(vec![]));
        serde_json::from_value(results).map_err(|err| err.to_string())
    }

    fn sync_collection<T: serde::Serialize>(&self, endpoint: &str, values: &[T]) -> Result<(), String> {
        for value in values {
            let body = serde_json::to_value(value).map_err(|err| err.to_string())?;
            self.request_value(endpoint, Method::POST, Some(body))?;
        }
        Ok(())
    }

    fn request_value(
        &self,
        endpoint: &str,
        method: Method,
        data: Option<Value>,
    ) -> Result<Value, String> {
        let url = format!("{}{}", self.api_url, endpoint);
        let mut request = self.client.request(method, &url).header("Content-Type", "application/json");

        if let Some(key) = &self.api_key {
            request = request.header("X-API-Key", key);
        }

        if let Some(payload) = data {
            request = request.json(&payload);
        }

        let response = request.send().map_err(|err| err.to_string())?;
        let status = response.status();
        let text = response.text().map_err(|err| err.to_string())?;

        let parsed: Value = serde_json::from_str(&text).map_err(|_| format!("Failed to parse response: {}", text))?;
        if status.is_success() {
            Ok(parsed)
        } else {
            let message = parsed
                .get("error")
                .and_then(|value| value.as_str())
                .unwrap_or(&text)
                .to_string();
            Err(format!("HTTP {}: {}", status.as_u16(), message))
        }
    }
}
